platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/Tipset.java

```diff
@@ -0,0 +1,199 @@
+        }
+
+        final int length = tipsets.get(0).tips.length;
+        final Tipset newTipset = buildEmptyTipset(tipsets.get(0));
+
+        for (int index = 0; index < length; index++) {
+            long max = -1;
+            for (final Tipset tipSet : tipsets) {
+                max = Math.max(max, tipSet.tips[index]);
+            }
+            newTipset.tips[index] = max;
+        }
+
+        return newTipset;
+    }
+
+    /**
+     * Get the tip generation for a given node ID.
+     *
+     * @param nodeId the node ID in question
+     * @return the tip generation for the node ID
+     */
+    public long getTipGenerationForNodeId(@NonNull final NodeId nodeId) {
+        final int index = addressBook.getIndexOfNodeId(nodeId);
+        return tips[index];
+    }
+
+    /**
+     * Get the tip generation for a given node
+     *
+     * @param nodeId the node in question
+     * @return the tip generation for the node
+     */
+    public long getTipGenerationForNode(@NonNull final NodeId nodeId) {
+        return tips[addressBook.getIndexOfNodeId(nodeId)];
+    }
+
+    /**
+     * Get the number of tips currently being tracked.
+     *
+     * @return the number of tips
+     */
+    public int size() {
+        return tips.length;
+    }
+
+    /**
+     * Advance a single tip within the tipset.
+     *
+     * @param creator    the node ID of the creator of the event
+     * @param generation the generation of the event
+     * @return this object
+     */
+    public @NonNull Tipset advance(@NonNull final NodeId creator, final long generation) {
+        final int index = addressBook.getIndexOfNodeId(creator);
+        tips[index] = Math.max(tips[index], generation);
+        return this;
+    }
+
+    /**
+     * <p>
+     * Get the combined weight of all nodes which experienced a tip advancement between this tipset and another tipset.
+     * Note that this method ignores advancement contributions from this node.
+     * </p>
+     *
+     * <p>
+     * A tip advancement is defined as an increase in the tip generation for a node ID. The exception to this rule is
+     * that an increase in generation for the target node ID is never counted as a tip advancement. The tip advancement
+     * count is defined as the sum of all tip advancements after being appropriately weighted.
+     * </p>
+     *
+     * <p>
+     * Advancements of non-zero stake nodes are tracked via {@link TipsetAdvancementWeight#advancementWeight()}, while
+     * advancements of zero stake nodes are tracked via {@link TipsetAdvancementWeight#zeroStakeAdvancementWeight()}.
+     *
+     * @param selfId compute the advancement count relative to this node ID
+     * @param that   the tipset to compare to
+     * @return the tipset advancement weight
+     */
+    @NonNull
+    public TipsetAdvancementWeight getTipAdvancementWeight(@NonNull final NodeId selfId, @NonNull final Tipset that) {
+        long nonzeroCount = 0;
+        long zeroCount = 0;
```

Comment:
This is only a count because each zero-weighted node is being given a pretend weight of 1. I think it's clearer to name these `nonzeroWeightSum` and `zeroWeightSum`

Reply:
I've renamed these variables:

```
        long nonZeroWeight = 0;
        long zeroWeightCount = 0;
```